feat(frontend): Add unit tests

stevefaeembra opened the pull request:
Add unit tests, fixed existing test failures

stevefaeembra committed:
feat(frontend): Add unit tests

Add unit tests, fixed existing test failures

stevefaeembra committed:
fix(frontend): Add vs code settings

vscode settings were not in source code control, fixed this

marksmall commented on client/setupTests.ts:
```diff
@@ -1,10 +1,23 @@
+// @ts-ignore
+//global.Request = Request;
+///@ts-ignore
+//global.Response = Response;
```
This commented out code should be removed, it was an attempt at a bandage, but was ultimately not needed. I don't like adding eslint-disable lines, I get why you did it, but they shouldn't just be added to make our lives easier by ignoring the underlying reason for the warning in the first place.

marksmall commented on client/setupTests.ts:
```diff
@@ -1,10 +1,23 @@
+
+beforeAll(() => {
+  server.listen({ onUnhandledRequest: 'error' });
+});
```
This should be reverted back, no need for curly braces on the anon function.

marksmall commented on client/src/accounts/authentication/authentication.hook.test.tsx:
```diff
@@ -0,0 +1,39 @@
+import { AuthenticationProvider } from './authentication.context';
+import { useAuthentication } from './authentication.hook';
+
+interface Foo {
```
Please stop using silly terms like Foo, it gets really frustrating reading the code. I removed all the content of this interface and it made no difference, suggesting it isn't absolutely necessary. The only test using it is actually `skipped`, so it is never really used. The other test also uses `only`, so you have 2 tests, one skipped and if you added any others, they wouldn't run as this one is marked as the only one that should run.

I'm not sure the first test is of any real use, your actually testing react functionality, not the context/provider.

The test marked as to run doesn't use any generics, you should use the one you define in the test, along with the type returned, you can get that from the `authentication.context.tsx` file, it's called `AuthenticationContextType`.

You rerender the hook, I don't know why, that might explain the `result.result`, but I'm not exactly sure on this.

Basically, there are many mistakes in this that it is not ready to review further. Again, make a list of tests and circulate them if your not sure.

marksmall commented on client/src/accounts/authentication/login.hook.test.tsx:
```diff
@@ -0,0 +1,67 @@
+describe('useLoginHook', () => {
+  it('should throw an error for an unknown user', async () => {
+    const loginData = { email: 'bob@example.com', password: 'otherpassword', accessToken: 'foobar' };
+    const { result } = renderHook<Result, unknown>(() => useLogin());
```
You only seem to use generics on the first test, this is wrong, now maybe we should define the result inside the hook being tested, just like we define the input e.g. `LoginFormType`

marksmall commented on client/src/accounts/authentication/login.hook.test.tsx:
```diff
@@ -0,0 +1,67 @@
+  it('should throw an error for an unknown user', async () => {
+    const loginData = { email: 'bob@example.com', password: 'otherpassword', accessToken: 'foobar' };
+    const { result } = renderHook<Result, unknown>(() => useLogin());
+    server.use(rest.post(ENDPOINT, (req, res, ctx) => res(ctx.status(400), ctx.json({ message: 'User not found' }))));
```
It would be good to start using constants for HTTP statuses. I don't know about anyone else but I constantly have to look them up. I did this for the server, see `server/src/utils/http.ts`, this should also exist in the client-side.

marksmall commented on client/src/accounts/authentication/login.hook.test.tsx:
```diff
@@ -0,0 +1,67 @@
+  mutate: (form: Record<string, unknown>) => void;
+  error: Record<string, unknown>;
+  current: Record<string, unknown>;
+}
```
We should look to see if react-query provides better types for the likes of `mutate|error`, if not, we should create our own instead of defining them over and over, in separate tests. I'm sure you've already likely duplicated this elsewhere.

marksmall commented on client/src/accounts/authentication/login.hook.test.tsx:
Please use whitespace to demarcate things, all your tests run into each other making it hard to see where one test ends and another starts. Same with code inside, separate renderHook, from the act that mutates and separate that from the expects. I find your code really hard to read at times.

marksmall commented on client/src/accounts/authentication/login.hook.test.tsx:
```diff
@@ -0,0 +1,67 @@
+    act(() => {
+      result.current.mutate(loginData);
+    });
+    await waitFor(() => {
```
This doesn't really follow the pattern I stated in slack 2 days ago. The maintainers of react-query, at least I think it was them, suggest awaiting the `result.current.isSuccess`, then after that, the expect lines can just be synchronous and nothing needs to be or should be returned from a test or an await block. If we follow a common pattern, then everyone's life is made easier.

marksmall commented on client/src/accounts/authentication/login.hook.ts:
```diff
@@ -30,36 +30,50 @@ export const useLogin = () => {
+      //console.log('** In the Mutator, big whoop! **');
+      //console.log('LOGIN FORM: ', form);
+      //console.log('ENDPOINT:', ENDPOINT);
+      const response = await apiClient(ENDPOINT, form);
```
I think this should be reverted back to `data`, `response` suggests it is a Response object, but I don't believe that is true.

marksmall commented on client/src/accounts/authentication/register.hook.test.tsx:
```diff
@@ -0,0 +1,107 @@
+      lastName: 'Ross',
+      password: 'otherpassword',
+    };
+    const { result } = renderHook<Result, unknown>(() => useRegister());
```
Missing type for the form itself, this can be retrieved from the hook and your `registerForm` should adhere to it. I know it is a headache, but we gotta start writing TypeScript, not JavaScript.

marksmall commented on client/src/accounts/authentication/register.hook.test.tsx:
```diff
@@ -0,0 +1,107 @@
+    act(() => {
+      result.current.mutate(registerForm);
+    });
+    await waitFor(() => {
```
Try awaiting the `result.current.isSuccess` then after that, do expect checks synchronously. That is something I failed to do in my error report hook test, I only learned about that after writing that test.

marksmall commented on client/src/api/api-client.hook.ts:
```diff
@@ -67,9 +68,11 @@ export const useApiClient = () => {
+      // if (!response.ok) {
+      //   console.log('response in apiClient', response);
+      //   return Promise.reject({ message: 'Please re-authenticate' });
+      // }
```
Why have you commented this out? I'm not saying your wrong, just trying to understand why.

marksmall reviewed (commented):
Got a few comments, I think you've tried to do too much in this, your testing hooks and fixing app components, that is too wide a focus, hence 21 files are touched. You also remove a lot of console.logs, which I completely understand why, they can definitely get in the way when running tests.

You don't seem to be doing the same things consistently though, I saw a few test files that the first test used a generic response from the hook, but the others didn't, some skipped tests, which isn't terrible, but leaving tests marked as `only` is bad.

On the whole a decent start, but try to narrow your focus to one thing and one thing only please.

stevefaeembra committed:
feat(frontend): WIP - additional tests

Additional tests. ignore this branch, its a WIP
These will be split into individual PRs shortly
